File: python/pacer/sim/throttle.py
"""Probabilistic throttling layer.

Given a participation probability p from the controller, decide participation per
request with a single PRNG draw: participate iff U[0,1) < p. We use numpy's
Generator (PCG64) — a fast, statistically sound, NON-cryptographic PRNG. Crypto
randomness would be pointless overhead on the hot path.

The Go serving layer mirrors this with its own fast PRNG (math/rand, not
crypto/rand); the *statistical* properties are what must match, not the exact
stream.
"""

from __future__ import annotations

import numpy as np


class UniformThrottle:
    """Each request participates independently with probability p."""

    def __init__(self, seed: int = 0):
        self.rng = np.random.default_rng(seed)

    def participate(self, p: float) -> bool:
        if p >= 1.0:
            return True
        if p <= 0.0:
            return False
        return bool(self.rng.random() < p)

    def participate_batch(self, p: float, n: int) -> np.ndarray:
        """Vectorized draws — used by tests and offline analysis."""
        if p >= 1.0:
            return np.ones(n, dtype=bool)
        if p <= 0.0:
            return np.zeros(n, dtype=bool)
        return self.rng.random(n) < p

File: go/internal/throttle/throttle.go
// Package throttle implements the hot-path participation decision.
//
// Given a participation probability p from the controller, a request participates
// iff a fast PRNG draw is below p. We use math/rand/v2 (PCG), NOT crypto/rand:
// this is on the critical path and cryptographic randomness would be pointless
// overhead. This mirrors the Python UniformThrottle; the statistical properties
// match even though the exact streams differ.
package throttle

import "math/rand/v2"

// Throttle holds a fast PRNG. Not safe for concurrent use by multiple goroutines;
// give each worker its own, or guard it.
type Throttle struct {
	rng *rand.Rand
}

// New seeds a throttle deterministically.
func New(seed uint64) *Throttle {
	return &Throttle{rng: rand.New(rand.NewPCG(seed, seed^0x9E3779B97F4A7C15))}
}

// Participate returns true with probability p. p<=0 never participates, p>=1
// always participates.
func (t *Throttle) Participate(p float64) bool {
	if p <= 0 {
		return false
	}
	if p >= 1 {
		return true
	}
	return t.rng.Float64() < p
}
